File: backend-go/internal/modules/system/service/job_service.go
package service

import (
	"context"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"drip-admin/backend-go/internal/common"
	"github.com/gin-gonic/gin"
)

func (s *Server) Jobs(c *gin.Context) {
	page, err := common.ParsePage(c)
	if err != nil {
		common.HandleError(c, err)
		return
	}
	status, err := parseOptionalInt(c, "status")
	if err != nil {
		common.HandleError(c, err)
		return
	}
	query := s.db.Model(&SysJob{}).Where("deleted = 0")
	query = likeIfPresent(query, "job_name", c.Query("jobName"))
	query = likeIfPresent(query, "remark", c.Query("remark"))
	query = eqIfPresent(query, "status", status)
	query = likeIfPresent(query, "created_at", c.Query("createdAt")).Order("created_at desc")
	var rows []SysJob
	total, err := s.pageResult(query, page, &rows)
	successOrError(c, common.PageResult[SysJob]{List: rows, Total: total, Page: page.Page, PageSize: page.PageSize}, err)
}

func (s *Server) Job(c *gin.Context) {
	id, err := idParam(c)
	if err != nil {
		common.HandleError(c, err)
		return
	}
	var row SysJob
	if err := s.db.Where("id = ? and deleted = 0", id.Int64()).First(&row).Error; err != nil {
		common.HandleError(c, common.NotFound("operation failed"))
		return
	}
	common.Success(c, row)
}

func (s *Server) JobScripts(c *gin.Context) {
	extensions, err := scriptExtensions(c.Query("executorType"))
	if err != nil {
		common.HandleError(c, err)
		return
	}
	entries, err := os.ReadDir(s.cfg.Job.ScriptDir)
	if err != nil {
		common.Success(c, []string{})
		return
	}
	files := []string{}
	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		ext := strings.ToLower(filepath.Ext(entry.Name()))
		if extensions[ext] {
			files = append(files, entry.Name())
		}
	}
	sort.Strings(files)
	common.Success(c, files)
}

func (s *Server) CreateJob(c *gin.Context) {
	var request JobSaveRequest
	if err := common.BindJSON(c, &request); err != nil {
		common.HandleError(c, err)
		return
	}
	id, err := s.saveJob(0, request, true)
	successOrError(c, id, err)
}

func (s *Server) UpdateJob(c *gin.Context) {
	id, err := idParam(c)
	if err != nil {
		common.HandleError(c, err)
		return
	}
	var request JobSaveRequest
	if err := common.BindJSON(c, &request); err != nil {
		common.HandleError(c, err)
		return
	}
	_, err = s.saveJob(id, request, false)
	successOrError(c, nil, err)
}

func (s *Server) DeleteJob(c *gin.Context) {
	id, err := idParam(c)
	if err != nil {
		common.HandleError(c, err)
		return
	}
	successOrError(c, nil, s.db.Model(&SysJob{}).Where("id = ?", id.Int64()).Update("deleted", 1).Error)
}

func (s *Server) JobStatus(c *gin.Context) {
	id, err := idParam(c)
	if err != nil {
		common.HandleError(c, err)
		return
	}
	status, err := statusRequest(c)
	if err != nil {
		common.HandleError(c, err)
		return
	}
	successOrError(c, nil, s.db.Model(&SysJob{}).Where("id = ? and deleted = 0", id.Int64()).Update("status", status).Error)
}

func (s *Server) RunJob(c *gin.Context) {
	id, err := idParam(c)
	if err != nil {
		common.HandleError(c, err)
		return
	}
	var job SysJob
	if err := s.db.Where("id = ? and deleted = 0", id.Int64()).First(&job).Error; err != nil {
		common.HandleError(c, common.NotFound("operation failed"))
		return
	}
	go s.executeJob(job)
	common.Success(c, nil)
}

func (s *Server) JobLogs(c *gin.Context) {
	id, err := idParam(c)
	if err != nil {
		common.HandleError(c, err)
		return
	}
	s.jobRunLogPage(c, &id)
}

func (s *Server) JobRunLogs(c *gin.Context) {
	s.jobRunLogPage(c, nil)
}

func (s *Server) jobRunLogPage(c *gin.Context, jobID *common.Int64String) {
	page, err := common.ParsePage(c)
	if err != nil {
		common.HandleError(c, err)
		return
	}
	query := s.db.Model(&SysJobRunLog{})
	if jobID != nil {
		query = query.Where("job_id = ?", jobID.Int64())
	}
	query = likeIfPresent(query, "job_name", c.Query("jobName"))
	query = likeIfPresent(query, "status", c.Query("status")).Order("started_at desc")
	var rows []SysJobRunLog
	total, err := s.pageResult(query, page, &rows)
	successOrError(c, common.PageResult[SysJobRunLog]{List: rows, Total: total, Page: page.Page, PageSize: page.PageSize}, err)
}

func (s *Server) saveJob(id common.Int64String, request JobSaveRequest, create bool) (common.Int64String, error) {
	if err := validateJob(request); err != nil {
		return 0, err
	}
	if create {
		row := SysJob{ID: common.NewID(), JobName: request.JobName, CronExpression: request.CronExpression, ExecutorType: request.ExecutorType, ScriptFile: optionalString(request.ScriptFile), ScriptArgs: optionalString(request.ScriptArgs), ClassName: optionalString(request.ClassName), MethodName: optionalString(request.MethodName), Status: intOrDefault(request.Status, 1), Remark: optionalString(request.Remark)}
		return row.ID, s.db.Create(&row).Error
	}
	return id, s.db.Model(&SysJob{}).Where("id = ? and deleted = 0", id.Int64()).Updates(map[string]any{
		"job_name":        request.JobName,
		"cron_expression": request.CronExpression,
		"executor_type":   request.ExecutorType,
		"script_file":     optionalString(request.ScriptFile),
		"script_args":     optionalString(request.ScriptArgs),
		"class_name":      optionalString(request.ClassName),
		"method_name":     optionalString(request.MethodName),
		"status":          intOrDefault(request.Status, 1),
		"remark":          optionalString(request.Remark),
	}).Error
}

func validateJob(request JobSaveRequest) error {
	if err := common.RequiredString(request.JobName, "jobName"); err != nil {
		return err
	}
	if err := common.RequiredString(request.CronExpression, "cronExpression"); err != nil {
		return err
	}
	if len(strings.Fields(request.CronExpression)) < 5 || len(request.CronExpression) > 64 {
		return common.NewBusinessError(400000, "cronExpression format is invalid")
	}
	if err := common.RequiredString(request.ExecutorType, "executorType"); err != nil {
		return err
	}
	if strings.EqualFold(request.ExecutorType, "java") {
		if err := common.RequiredString(request.ClassName, "className"); err != nil {
			return err
		}
		return common.RequiredString(request.MethodName, "methodName")
	}
	return common.RequiredString(request.ScriptFile, "scriptFile")
}

func scriptExtensions(executorType string) (map[string]bool, error) {
	switch strings.ToLower(strings.TrimSpace(executorType)) {
	case "shell":
		return map[string]bool{".sh": true}, nil
	case "bat":
		return map[string]bool{".bat": true, ".cmd": true}, nil
	case "powershell", "ps1":
		return map[string]bool{".ps1": true}, nil
	case "python":
		return map[string]bool{".py": true}, nil
	default:
		return nil, common.NewBusinessError(400000, "executorType is not supported")
	}
}

func (s *Server) executeJob(job SysJob) {
	started := time.Now()
	runLog := SysJobRunLog{ID: common.NewID(), JobID: job.ID, JobName: job.JobName, Status: "RUNNING", StartedAt: &started}
	_ = s.db.Create(&runLog).Error
	err := s.executeJobTarget(job)
	finished := time.Now()
	status := "SUCCESS"
	var message *string
	if err != nil {
		status = "FAIL"
		text := err.Error()
		if len(text) > 512 {
			text = text[:512]
		}
		message = &text
	}
	_ = s.db.Model(&SysJobRunLog{}).Where("id = ?", runLog.ID.Int64()).Updates(map[string]any{
		"status":        status,
		"finished_at":   finished,
		"cost_ms":       common.Int64String(finished.Sub(started).Milliseconds()),
		"error_message": message,
	}).Error
}

func (s *Server) executeJobTarget(job SysJob) error {
	executorType := strings.ToLower(strings.TrimSpace(job.ExecutorType))
	if executorType == "java" {
		return nil
	}
	if job.ScriptFile == nil || strings.TrimSpace(*job.ScriptFile) == "" {
		return common.NewBusinessError(400000, "scriptFile is required")
	}
	script := filepath.Clean(filepath.Join(s.cfg.Job.ScriptDir, *job.ScriptFile))
	root, _ := filepath.Abs(s.cfg.Job.ScriptDir)
	full, _ := filepath.Abs(script)
	if !strings.HasPrefix(full, root) {
		return common.NewBusinessError(400000, "script path is not allowed")
	}
	command, err := jobCommand(executorType, full, common.StringValue(job.ScriptArgs))
	if err != nil {
		return err
	}
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Minute)
	defer cancel()
	return exec.CommandContext(ctx, command[0], command[1:]...).Run()
}

func jobCommand(executorType string, script string, args string) ([]string, error) {
	command := []string{}
	switch executorType {
	case "shell":
		command = append(command, "bash", script)
	case "bat":
		command = append(command, "cmd.exe", "/c", script)
	case "powershell", "ps1":
		command = append(command, "powershell.exe", "-ExecutionPolicy", "Bypass", "-File", script)
	case "python":
		command = append(command, "python", script)
	default:
		return nil, common.NewBusinessError(400000, "executorType is not supported")
	}
	command = append(command, strings.Fields(args)...)
	return command, nil
}

File: backend-rust/src/modules/system/service/job_service.rs
use crate::common::{next_id, AppError, I64String, PageParams, PageResult};
use crate::config::Settings;
use crate::modules::system::dto::job_request::JobSaveRequest;
use crate::modules::system::entity::sys_job::SysJob;
use crate::modules::system::entity::sys_job_run_log::SysJobRunLog;
use rbatis::RBatis;
use serde::Deserialize;
use std::path::PathBuf;
use std::sync::Arc;
use std::time::{Duration, Instant};

pub async fn list(
    database: Option<&Arc<RBatis>>,
    page: PageParams,
) -> Result<PageResult<SysJob>, AppError> {
    let database = require_database(database)?;
    let rows: Vec<SysJob> = database
        .exec_decode(
            "select id, job_name, cron_expression, executor_type, script_file, script_args, status, remark, deleted from sys_job where deleted = 0 order by created_at desc limit ? offset ?",
            vec![
                rbs::value!(page.page_size),
                rbs::value!(((page.page - 1) * page.page_size) as i64),
            ],
        )
        .await
        .map_err(map_database_error)?;
    let totals: Vec<Count> = database
        .exec_decode("select count(*) as total from sys_job where deleted = 0", vec![])
        .await
        .map_err(map_database_error)?;
    Ok(PageResult {
        list: rows,
        total: I64String(totals.first().map(|row| row.total).unwrap_or(0)),
        page: page.page,
        page_size: page.page_size,
    })
}

pub async fn scripts(settings: &Settings) -> Result<Vec<String>, AppError> {
    let root = script_root(settings)?;
    let mut directory = tokio::fs::read_dir(root)
        .await
        .map_err(|error| AppError::system(error.to_string()))?;
    let mut scripts = Vec::new();
    while let Some(entry) = directory
        .next_entry()
        .await
        .map_err(|error| AppError::system(error.to_string()))?
    {
        if entry
            .file_type()
            .await
            .map_err(|error| AppError::system(error.to_string()))?
            .is_file()
        {
            scripts.push(entry.file_name().to_string_lossy().to_string());
        }
    }
    scripts.sort();
    Ok(scripts)
}

pub async fn detail(database: Option<&Arc<RBatis>>, id: i64) -> Result<SysJob, AppError> {
    let rows: Vec<SysJob> = require_database(database)?
        .exec_decode(
            "select id, job_name, cron_expression, executor_type, script_file, script_args, status, remark, deleted from sys_job where id = ? and deleted = 0",
            vec![rbs::value!(id)],
        )
        .await
        .map_err(map_database_error)?;
    rows.into_iter()
        .next()
        .ok_or_else(|| AppError::not_found("资源不存在"))
}

pub async fn create(
    database: Option<&Arc<RBatis>>,
    request: JobSaveRequest,
) -> Result<I64String, AppError> {
    validate(&request)?;
    let id = next_id();
    require_database(database)?
        .exec(
            "insert into sys_job(id, job_name, cron_expression, executor_type, script_file, script_args, status, remark, deleted) values (?, ?, ?, ?, ?, ?, ?, ?, 0)",
            vec![
                rbs::value!(id),
                rbs::value!(request.job_name),
                rbs::value!(request.cron_expression),
                rbs::value!(request.executor_type.trim().to_ascii_lowercase()),
                rbs::value!(request.script_file),
                rbs::value!(request.script_args),
                rbs::value!(request.status.unwrap_or(1)),
                rbs::value!(request.remark),
            ],
        )
        .await
        .map_err(map_database_error)?;
    Ok(I64String(id))
}

pub async fn update(
    database: Option<&Arc<RBatis>>,
    id: i64,
    request: JobSaveRequest,
) -> Result<(), AppError> {
    detail(database, id).await?;
    validate(&request)?;
    require_database(database)?
        .exec(
            "update sys_job set job_name = ?, cron_expression = ?, executor_type = ?, script_file = ?, script_args = ?, status = ?, remark = ? where id = ?",
            vec![
                rbs::value!(request.job_name),
                rbs::value!(request.cron_expression),
                rbs::value!(request.executor_type.trim().to_ascii_lowercase()),
                rbs::value!(request.script_file),
                rbs::value!(request.script_args),
                rbs::value!(request.status.unwrap_or(1)),
                rbs::value!(request.remark),
                rbs::value!(id),
            ],
        )
        .await
        .map_err(map_database_error)?;
    Ok(())
}

pub async fn delete(database: Option<&Arc<RBatis>>, id: i64) -> Result<(), AppError> {
    detail(database, id).await?;
    require_database(database)?
        .exec("update sys_job set deleted = 1 where id = ?", vec![rbs::value!(id)])
        .await
        .map_err(map_database_error)?;
    Ok(())
}

pub async fn status(database: Option<&Arc<RBatis>>, id: i64, status: i32) -> Result<(), AppError> {
    if !matches!(status, 0 | 1) {
        return Err(AppError::bad_request("status is invalid"));
    }
    detail(database, id).await?;
    require_database(database)?
        .exec(
            "update sys_job set status = ? where id = ?",
            vec![rbs::value!(status), rbs::value!(id)],
        )
        .await
        .map_err(map_database_error)?;
    Ok(())
}

pub async fn run(database: Arc<RBatis>, settings: Arc<Settings>, id: i64) -> Result<(), AppError> {
    let job = detail(Some(&database), id).await?;
    let log_id = next_id();
    database
        .exec(
            "insert into sys_job_run_log(id, job_id, job_name, status, started_at, cost_ms) values (?, ?, ?, 'RUNNING', now(), 0)",
            vec![rbs::value!(log_id), rbs::value!(id), rbs::value!(&job.job_name)],
        )
        .await
        .map_err(map_database_error)?;
    tokio::spawn(async move {
        let started = Instant::now();
        let result = execute(&job, &settings).await;
        let (status, error_message) = match result {
            Ok(()) => ("SUCCESS", None),
            Err(error) => ("FAIL", Some(error.message.chars().take(512).collect::<String>())),
        };
        let _ = database
            .exec(
                "update sys_job_run_log set status = ?, finished_at = now(), cost_ms = ?, error_message = ? where id = ?",
                vec![
                    rbs::value!(status),
                    rbs::value!(started.elapsed().as_millis() as i64),
                    rbs::value!(error_message),
                    rbs::value!(log_id),
                ],
            )
            .await;
    });
    Ok(())
}

pub async fn run_logs(
    database: Option<&Arc<RBatis>>,
    page: PageParams,
    job_id: Option<i64>,
) -> Result<PageResult<SysJobRunLog>, AppError> {
    let database = require_database(database)?;
    let where_sql = if job_id.is_some() { "where job_id = ?" } else { "" };
    let mut args = Vec::new();
    if let Some(job_id) = job_id {
        args.push(rbs::value!(job_id));
    }
    let count_args = args.clone();
    args.push(rbs::value!(page.page_size));
    args.push(rbs::value!(((page.page - 1) * page.page_size) as i64));
    let rows: Vec<SysJobRunLog> = database
        .exec_decode(
            &format!("select id, job_id, job_name, status, started_at, finished_at, cost_ms, error_message from sys_job_run_log {where_sql} order by started_at desc, id desc limit ? offset ?"),
            args,
        )
        .await
        .map_err(map_database_error)?;
    let totals: Vec<Count> = database
        .exec_decode(
            &format!("select count(*) as total from sys_job_run_log {where_sql}"),
            count_args,
        )
        .await
        .map_err(map_database_error)?;
    Ok(PageResult {
        list: rows,
        total: I64String(totals.first().map(|row| row.total).unwrap_or(0)),
        page: page.page,
        page_size: page.page_size,
    })
}

async fn execute(job: &SysJob, settings: &Settings) -> Result<(), AppError> {
    let script = job
        .script_file
        .as_deref()
        .ok_or_else(|| AppError::bad_request("scriptFile is required"))?;
    let root = script_root(settings)?;
    let path = root
        .join(script)
        .canonicalize()
        .map_err(|_| AppError::bad_request("script path is not allowed"))?;
    if !path.starts_with(&root) {
        return Err(AppError::bad_request("script path is not allowed"));
    }
    let mut command = match job.executor_type.as_str() {
        "shell" => {
            let mut command = tokio::process::Command::new("bash");
            command.arg(&path);
            command
        }
        "bat" => {
            let mut command = tokio::process::Command::new("cmd.exe");
            command.args(["/c", path.to_string_lossy().as_ref()]);
            command
        }
        "powershell" | "ps1" => {
            let mut command = tokio::process::Command::new("powershell.exe");
            command.args(["-ExecutionPolicy", "Bypass", "-File", path.to_string_lossy().as_ref()]);
            command
        }
        "python" => {
            let mut command = tokio::process::Command::new("python");
            command.arg(&path);
            command
        }
        _ => return Err(AppError::bad_request("executorType is not supported")),
    };
    command.current_dir(&root);
    if let Some(args) = job.script_args.as_deref() {
        command.args(args.split_whitespace());
    }
    let result = tokio::time::timeout(Duration::from_secs(1800), command.status())
        .await
        .map_err(|_| AppError::system("script execution timed out"))?
        .map_err(|error| AppError::system(error.to_string()))?;
    if result.success() {
        Ok(())
    } else {
        Err(AppError::system(format!(
            "script execution failed, exit code: {}",
            result.code().unwrap_or(-1)
        )))
    }
}

fn script_root(settings: &Settings) -> Result<PathBuf, AppError> {
    PathBuf::from(&settings.job.script_dir)
        .canonicalize()
        .map_err(|_| AppError::system("scripts directory is not configured"))
}

fn require_database(database: Option<&Arc<RBatis>>) -> Result<&RBatis, AppError> {
    database
        .map(AsRef::as_ref)
        .ok_or_else(|| AppError::system("Rbatis database is not configured"))
}

fn validate(request: &JobSaveRequest) -> Result<(), AppError> {
    if request.job_name.trim().is_empty()
        || request.cron_expression.split_whitespace().count() < 5
        || request.cron_expression.len() > 64
    {
        return Err(AppError::bad_request("jobName or cronExpression is invalid"));
    }
    if !matches!(
        request.executor_type.trim().to_ascii_lowercase().as_str(),
        "shell" | "bat" | "powershell" | "ps1" | "python"
    ) {
        return Err(AppError::bad_request("executorType is not supported"));
    }
    if request.script_file.trim().is_empty() {
        return Err(AppError::bad_request("scriptFile is required"));
    }
    Ok(())
}

fn map_database_error(error: rbatis::Error) -> AppError {
    AppError::system(error.to_string())
}

#[derive(Deserialize)]
struct Count {
    total: i64,
}
